pandaproxy/sr: Fix normalized rendering for custom options

IoannisRP opened the pull request:
When calling CopyTo from `FileDescriptor` to `FileDescriptorProto`, custom options are discarded.

To fix it a manual pass has been added to copy the custom options from the old `FileDescriptorProto` to the new `FileDescriptorProto`.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.3.x
- [x] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

IoannisRP committed:
pandaproxy/sr: fix protobuf options rendering

When an option is made of multiple names, the names need to
be separated by a ".".

IoannisRP committed:
pandaproxy/sr: add uninterpreted options rendering support for oneof fields

IoannisRP committed:
pandaproxy/sr: fix normalized support for uninterpreted options

When calling CopyTo, uninterpreted options are discarded. A manual
pass to copy them from the old FileDescriptorProto has been added.

IoannisRP commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -131,18 +131,18 @@ struct fmt::formatter<google::protobuf::UninterpretedOption>
-                  ctx.out(), "{} = \"{}\"", fmt::join(option.name(), ""), val);
+                  ctx.out(), "{} = \"{}\"", fmt::join(option.name(), "."), val);
             } else if (option.has_aggregate_value()) {
                 return fmt::format_to(
                   ctx.out(),
                   "{} = {{{}\n{:{}}}}",
-                  fmt::join(option.name(), ""),
+                  fmt::join(option.name(), "."),
                   val,
                   "",
                   indent + 2);
             }
             return fmt::format_to(
-              ctx.out(), "{} = {}", fmt::join(option.name(), ""), val);
+              ctx.out(), "{} = {}", fmt::join(option.name(), "."), val);
```
without these changes, options like
```
option (my_method_option).foo = 567;
```
were being rendered as
```
option (my_method_option)foo = 567;
```

IoannisRP commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -1046,7 +1152,15 @@ struct protobuf_schema_definition::impl {
             fmt::print(
               os, "{:{}}{} = {}", "", indent + 2, value.name(), value.number());
-            if (value.has_options()) {
+            const bool has_options = [&value]() {
+                if (!value.has_options()) {
+                    return false;
+                }
+                const auto& opts = value.options();
+                return opts.has_deprecated() || opts.has_debug_redact()
+                       || !opts.uninterpreted_option().empty();
+            }();
+            if (has_options) {
```
here and in service, after the copy was introduced, the `has_options` call would report `true` but without any of the internal conditionals being triggered.

Only these two exhibit this behavior (in the tests that we have).
This is a workaround till we understand why they report a "false" positive.

BenPope replied:
Do we need something like this?
```c++
    template<typename Proto>
    void copy_options(const Proto& from, Proto& to) const {
        if constexpr (requires { from.has_options(); }) {
            if (!from.has_options()) {
                return;
            }
        }
        to.mutable_options()->CopyFrom(from.options());
    }
```

IoannisRP replied:
what is the purpose of the `constexpr if` here? If it doesn't have it, it is a compile time error. But this is called on all the protos that we already query for options.

BenPope replied:
> what is the purpose of the `constexpr if` here? If it doesn't have it, it is a compile time error. But this is called on all the protos that we already query for options.

Oh, I thought I saw some types that didn't have an `has_options()`

IoannisRP replied:
```
            if (!from.has_options()) {
                return;
            }
```
this is the section that protects from this, right?
All Proto files should have a `has_options` function, no?

IoannisRP replied:
Dropped the extra checks. They are indeed not needed after adding the above check.

IoannisRP commented on src/v/pandaproxy/schema_registry/test/compatibility_protobuf.cc:
```diff
@@ -597,6 +597,384 @@ message Bar {
+}
+enum MyEnum {
+  VALUE_0 = 0;
+  VALUE_1 = 1 [(metadata) = {some_bool : true , some_string : "test_string"
+  }];
```
notice the `\n` before the `}]`.

Could this be potentially problematic?

BenPope replied:
> notice the `\n` before the `}]`.
> 
> Could this be potentially problematic?

I don't think so, as long as it's rendered consistently when normalized.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--e094cb44-1285-4298-b3e3-05a07d67bfe7-->
<!--TEST RESULTS-->
<!--eead35ff-3631-4ca3-a79c-1618a8c3174c-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60915">build#60915</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60915#0194759d-a891-401f-8fea-87bf20026afb | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60994">build#60994</a></summary>


| test_id                                                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_chunk_read_path_test.CloudStorageChunkReadTest.test_read_when_segment_size_smaller_than_chunk_size | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60994#019489ad-a88c-499c-a222-924c8d0c4348 | FLAKY         | 1/2      |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60994#019489ad-a88b-4ac3-9084-3b783a266bd9 | FLAKY         | 1/2      |
</details>
<!--/body-->

BenPope commented on src/v/pandaproxy/schema_registry/test/compatibility_protobuf.cc:
```diff
@@ -597,6 +597,384 @@ message Bar {
+  enum NestedEnum {
+    option (my_enum_option) = true;
+
+    FOO = 1 [(my_enum_value_option) = 432];
```
What happens if we have two custom options in here, can they be sorted and with canonical names under normalization?

BenPope reviewed (commented):
Unfortunately I don't think the copy works well for normalization.

IoannisRP committed:
pandaproxy/sr: fix normalization order by name for all custom options

Uninterpreted options (custom options) should be sorted by name when
normalized for all types of uninterpreted options. Added missing
sorting for enum options and enumValue options.

IoannisRP committed:
pandaproxy/sr: sort extentions by type/number

dotnwat commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -131,18 +131,18 @@ struct fmt::formatter<google::protobuf::UninterpretedOption>
             if (option.has_string_value()) {
                 return fmt::format_to(
-                  ctx.out(), "{} = \"{}\"", fmt::join(option.name(), ""), val);
+                  ctx.out(), "{} = \"{}\"", fmt::join(option.name(), "."), val);
```
>pandaproxy/sr: fix protobuf service options rendering

commit message is missing an explanation on why its broken and how this commit fixes it.

dotnwat commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -599,6 +600,102 @@ struct protobuf_schema_definition::impl {
+        for (int i_service = 0; i_service < n_services; ++i_service) {
+            const auto& service_from = from.service(i_service);
+            auto* service_to = to.mutable_service(i_service);
+            copy_uninterpreted_options(service_from, *service_to);
```
this is a recursive call here, right? historically we have had some cases where recursion depth has been an issue with the stack size in seastar. i don't know if that is only relevant for seastar::async context, or just generally true on the reactor. something to consider.

IoannisRP replied:
The depth of the recursion in the `copy_uninterpreted_options` will be as deep as the depth in the `render_proto`. And it is the depth of the proto graph. The whole architecture of this module should be revisited. When we get to it, that's something to keep in mind!

IoannisRP committed:
pandaproxy/sr: merge consecutive extensions to the same type

IoannisRP commented on src/v/pandaproxy/schema_registry/test/compatibility_protobuf.cc:
```diff
@@ -597,6 +597,442 @@ message Bar {
+  FOO = 1 [(my_enum_value_option_b) = 432, (my_enum_value_option_a) = 321];
+  BAR = 2;
+}
+extend google.protobuf.FileOptions {
+  optional string my_file_option_b = 50008;
+  optional string my_file_option_a = 50000;
+}
+extend google.protobuf.MessageOptions {
+  optional int32 my_message_option_b = 50009;
+  optional int32 my_message_option_a = 50001;
+}
+extend google.protobuf.FileOptions {
+  optional string my_file_option_c = 50015;
+}
```
in `sanitize` only extensions defined in order will be merged. If they are defined out of order, they will be rendered out of order.

IoannisRP committed:
pandaproxy/sr: fix normalization order for repeated custom options

- grouped all the duplicated calls to maybe_sort custom options into
a single helper function
- changed projection function from just the name of the custom option
to the full serialization of it. This way, both the name and the
value contributes to the ordering.

BenPope commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -932,6 +932,15 @@ struct protobuf_schema_definition::impl {
+              decl.options().uninterpreted_option(),
+              std::less{},
+              [](const pb::UninterpretedOption& o) {
+                  return fmt::format("{}", fmt::join(o.name(), ""));
```
Should this be joined with a '.' for consistency?

IoannisRP replied:
It doesn't make a difference, but yes, it should have been (i missed it).
It all got removed in https://github.com/redpanda-data/redpanda/pull/24856/commits/f0e56f474472c98ee6cb07f9fb4a554a3dd7a869 where we moved to `fmt::format("{}", o)` more a sorting that is value-aware.

IoannisRP replied:
Fixed

BenPope commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -599,6 +600,102 @@ struct protobuf_schema_definition::impl {
+            const auto& nested_from = from.nested_type(i_nested);
+            auto* nested_to = to.mutable_nested_type(i_nested);
+            copy_uninterpreted_options(nested_from, *nested_to);
+        }
```
I find these looks syntactically dense, perhaps something like:
```suggestion
        for (auto&& [nested_from, nested_to] :
             boost::combine(from.nested_type(), *to.mutable_nested_type())) {
            copy_uninterpreted_options(nested_from, nested_to);
        }
```

BenPope commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -599,6 +600,102 @@ struct protobuf_schema_definition::impl {
+
+            if (field_from.has_oneof_index()) {
+                const auto i_oneof = field_from.oneof_index();
+                const auto& oneof_from = from.oneof_decl(i_oneof);
+                auto* oneof_to = to.mutable_oneof_decl(i_oneof);
+                copy_options(oneof_from.options(), oneof_to->mutable_options());
```
Does naming the variables make it easier or harder to read?
```suggestion
                copy_options(
                  from.oneof_decl(i_oneof).options(),
                  to.mutable_oneof_decl(i_oneof)->mutable_options());
```

IoannisRP replied:
Personally, I find the extra variable easier on the eyes.

IoannisRP replied:
However, after the other changes, it becomes
```suggestion
                const auto i_oneof = field_from.oneof_index();
                copy_options(
                  from.oneof_decl(i_oneof), *to.mutable_oneof_decl(i_oneof));
```
which is pretty concise :ok_hand:

BenPope commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -913,13 +913,32 @@ struct protobuf_schema_definition::impl {
               return std::make_pair(extension.extendee(), extension.number());
           });
+
+        const auto start_section = [indent, &os](const std::string& ext) {
```
Proto are moving to string views in general, so to avoid string creation if that happens in a later version
```suggestion
        const auto start_section = [indent, &os](std::string_view ext) {
```

BenPope reviewed (commented):
Very cool

BenPope commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -599,6 +601,89 @@ struct protobuf_schema_definition::impl {
+        for (auto&& [field_from, field_to] :
+             boost::combine(from.field(), *to.mutable_field())) {
+            copy_uninterpreted_options(field_from, field_to);
+
+            if (field_from.has_oneof_index()) {
+                const auto i_oneof = field_from.oneof_index();
+                copy_options(
+                  from.oneof_decl(i_oneof), *to.mutable_oneof_decl(i_oneof));
+            }
```
I don't think this indirection is desirable.
```suggestion
        }

        // oneof messages
        for (auto&& [oneof_from, oneof_to] :
             boost::combine(from.oneof_decl(), *to.mutable_oneof_decl())) {
            copy_options(oneof_from, oneof_to);
```

BenPope commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -599,6 +601,89 @@ struct protobuf_schema_definition::impl {
+            copy_uninterpreted_options(enum_from, enum_to);
+        }
+
+        for (auto&& [service_from, service_to] :
```
Are we  missing
```c++
        for (auto&& [ext_from, ext_to] :
             boost::combine(from.extension(), *to.mutable_extension())) {
            copy_uninterpreted_options(ext_from, ext_to);
        }
```

IoannisRP replied:
Nice catch

IoannisRP committed:
pandaproxy/sr: Add support for extention custom options

BenPope commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -901,13 +901,32 @@ struct protobuf_schema_definition::impl {
-            fmt::print(
-              os, "{:{}}extend {} {{\n", "", indent, extension.extendee());
+            const auto& extendee = extension.extendee();
+            if (active_extendee != extendee) {
+                active_extendee = extendee;
+                if (open_section) {
+                    close_section();
+                }
+                start_section(extendee);
+                open_section = true;
+            }
```
We should make a note to extract this algorithm in a followup PR.

IoannisRP replied:
yes, this should be a `group_by`.

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-24856-v24.2.x-716 remotes/upstream/v24.2.x
git cherry-pick -x 44da70e866 cca81edac0 b2785967bb 057ec1293d a46feb43cf f15e7e4216 0ec5585a31 f8c11ae7b7
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/12893703739).

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-24856-v24.3.x-157 remotes/upstream/v24.3.x
git cherry-pick -x 44da70e866 cca81edac0 b2785967bb 057ec1293d a46feb43cf f15e7e4216 0ec5585a31 f8c11ae7b7
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/12893703043).
